### ISSUE TITLE
Slow query with full index usage

### ISSUE THREAD

**Issue by Minstel** [arangodb/arangodb#19352]
## My Environment

* __ArangoDB Version__:        3.7.12
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:  ArangoDB Starter in Docker
* __Configuration__:               `ARANGODB_OVERRIDE_DETECTED_TOTAL_MEMORY`: 32G, `ARANGODB_OVERRIDE_DETECTED_NUMBER_OF_CORES`: 4
* __Infrastructure__:               own 
* __Operating System__:        Ubuntu 20.04
* __Total RAM in your machine__:   hundreds of GB
* __Disks in use__:        SSD
* __Used Package__:   Docker - official Docker library

## Component, Query & Data

__Affected feature__:
AQL query using web interface

__AQL query (if applicable)__:
```
FOR report IN r_reports
FILTER report.main_task_state == 'success'
SORT report.date DESC, report._key DESC

FOR edge0 IN r_sha256_reports_edges
FILTER edge0._from == CONCAT('r_reports/', report._key) AND edge0._to IN ['fc4623b113a1f603c0d9ad5f83130bd6de1c62b973be9892305132389c8588de']
         
LIMIT 2000
RETURN {id: report._key, hash: report.source_sha256, date: report.date}

```
__AQL explain and/or profile (if applicable)__:

Explain: 
```
Execution plan:
 Id   NodeType             Est.   Comment
  1   SingletonNode           1   * ROOT
 15   IndexNode         1532856     - FOR report IN r_reports   /* reverse persistent index scan, projections: `date`, `_key`, `source_sha256` */    
 14   IndexNode         1532856       - FOR edge0 IN r_sha256_reports_edges   /* persistent index scan, scan only */    
 11   LimitNode            2000         - LIMIT 0, 2000
 12   CalculationNode      2000         - LET #10 = { "id" : report.`_key`, "hash" : report.`source_sha256`, "date" : report.`date` }   /* simple expression */   /* collections used: report : r_reports */
 13   ReturnNode           2000         - RETURN #10

Indexes used:
 By   Name                                 Type         Collection               Unique   Sparse   Selectivity   Fields                                  Ranges
 15   r_reports_main_task_state_date_key   persistent   r_reports                false    false        99.99 %   [ `main_task_state`, `date`, `_key` ]   (report.`main_task_state` == "success")
 14   r_sha256_reports_edges_from_to       persistent   r_sha256_reports_edges   true     false       100.00 %   [ `_from`, `_to` ]                      ((edge0.`_from` == CONCAT("r_reports/", report.`_key`)) && (edge0.`_to` == "fc4623b113a1f603c0d9ad5f83130bd6de1c62b973be9892305132389c8588de"))

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   move-calculations-up-2
  4   move-filters-up-2
  5   use-indexes
  6   remove-filter-covered-by-index
  7   use-index-for-sort
  8   remove-unnecessary-calculations-2
  9   reduce-extraction-to-projection
```

Profiling:

![2023-06-30_03-51](https://github.com/arangodb/arangodb/assets/6154708/43e12393-692e-4c1a-970c-0284c7ea9d0e)

__Dataset__:
`r_reports` collection holds data that looks like this:
```
{
  "source_sha256": "8c99a829dcb8760614d26c8226a8ed84584f4e94423dbb3e69dee4025b75eafb",
  "date": 1688042558.385,
  "vt_rate": -1,
  "user": "648b8a57b132e6a02c3ac4a2",
  "main_task_state": "success",
  "retry_count": 0,
  "is_source_archive": false,
  "source_type": "file"
}
```

`r_sha256_reports_edges` has records like this:
```
{
  "id": "r_sha256_reports_edges/1939",
  "_rev": "_gOaredm---"
  "_key": 1939
  "_from": "r_reports/4e97e378-dbc5-4258-a349-f9f6cb3423f7"
  "_to": "r_sha256/8c99a829dcb8760614d26c8226a8ed84584f4e94423dbb3e69dee4025b75eafb"
}
```

Each collection holds somewhere about a million records.

__Problem__:

Given query, when executed, takes 30-70 seconds. On profiling screen you can see that most of time takes looking though index in `r_sha256_reports_edges` collection (at least I understand profiling like this). But from index usage I'd tell that query should be execute instantly. 
What is weird is that we have similar queries with other collections joined to `r_reports` in the same way, that are executed very fast.

**Comment by dothebart** [comment]
Hi, 
Please note that ArangoDB 3.7 is EOL. Please upgrade to Arangodb 3.9 at least - better 3.10 or 3.11 and check back whether the issue persists.

**Comment by Minstel** [comment]
@dothebart ok, thank you. The hard thing here is the procedure needed for upgrade. Because application we have is also shipped to standalone customers. So we need some kind of script that will be able to do the update, with guarantee against issues. Maybe you have such script, that will perform docker arango upgrade?

**Comment by Minstel** [comment]
@dothebart Also, during indexes creation, it's possible that we had a loop of falling/restarting server workers. Though indexes were created with `background` option, is it possible that loop caused indexes to be created incorrectly? And thus database upgrade won't fix them

**Comment by dothebart** [comment]
creating indices in background operations if you i.e. have heavy write load is probably not the best Idea. But if you run OOM much may happen. Many bugs have been found and fixed since way back then. 

I would at least try to upgrade to the latest released 3.7 versions; for others I would test the application in advance. 

We use python scripts to test-upgrade starter-based installs here (if that may serve you as inspiration): https://github.com/arangodb/release-test-automation/ 
But thats not doing docker upgrade; our managed deployments over at ArangoGraph are managed using kubernetes. 
You need to be doing `--database.auto-upgrade true` launches inbetween before starting another version. 

Leaning on a script to do all that unattended doesn't sound like the best Idea to me.

**Comment by Minstel** [comment]
@dothebart Thanks)

Can you also clarify these words of yours:
> creating indices in background operations if you i.e. have heavy write load is probably not the best Idea

What can be the consequences?

**Comment by Minstel** [comment]
@dothebart We did a rebuild of indexes, hoping that it will fix issue, but it didn't. So I guess we'll need to do database upgrade.
Can you at least confirm that the query I provided to you should work fast? That it's correctly built?

**Comment by Minstel** [comment]
@dothebart We need this asap, so it it would be great if you give us feedback on query, for questions in my previous comment) Because the only purpose we use arango for is not working now.

**Comment by dothebart** [comment]
From the runtime of the individual nodes I'd say that you probably should play with where to set the `LIMIT` statement, or maybe have several ones.
